Turn on new migration plugin / import flow feature flag on wpcalypso

ouikhuan opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #77472 

## Proposed Changes

* This PR adds the new migration plugin flow feature flag to the `wpcalypso` env to prepare for the upcoming CfT.
* Based on the flow changes on the feature flag, we also needs to update our existing e2e tests.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

TeamCity test run against the live link of this PR

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

ouikhuan committed:
Turn on new migration plugin flow feature flag for wpcalypso

ouikhuan committed:
Update e2e importer flow test with new feature flag

bogiii reviewed (approved):
Looks good to me! 👍

worldomonation commented on packages/calypso-e2e/src/lib/flows/start-import-flow.ts:
```diff
@@ -14,6 +14,8 @@ const selectors = {
 	wpContentOnlyContinueButton:
 		'.content-chooser .import-layout__column:nth-child(2) > div > div:last-child button:text("Continue")',
+	wpPlansContentOnlyOptionButton:
+		'.import__import-everything--redesign > .import__footer-button-container:last-child button:text("Use the content-only import option")',
```
**comment:blocking** - in general, we try to avoid using CSS selectors like this, preferring to use [role-based selectors](https://playwright.dev/docs/release-notes#highlights-2). You can see this in action in most of the recent PRs authored by myself or others involved in the E2E maintenance effort.

ouikhuan replied:
Thanks, I've changed it to the role-based selector now.

worldomonation commented on packages/calypso-e2e/src/lib/flows/start-import-flow.ts:
```diff
@@ -114,6 +116,13 @@ export class StartImportFlow {
+	 * Validates that we've landed on the WordPress plans page.
+	 */
+	async validateWordPressPlansPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.wpPlansContentOnlyOptionButton );
```
**question:non-blocking**- without knowing the screen that this is referring to, I can't say for certain but the description sounds like it's checking that the user has been redirected to the`/plans` endpoint.

If so, can I suggest usage of the `PlansPage.validateActivePlan` or `PlansPage.validateActiveTab` in order to validate that the expected action took place _without_ introducing a new single-use method?

Alternatively, checking for `page.waitForURL` in the test is always a possibility.

ouikhuan replied:
Actually, it's a pre-migration page with plans information in it. I've renamed the function a little, and I hope it's clearer now.

worldomonation commented on packages/calypso-e2e/src/lib/flows/start-import-flow.ts:
```diff
@@ -128,6 +137,13 @@ export class StartImportFlow {
+	 * Continue 'content only' WordPress migration on plans page.
+	 */
+	async contentOnlyWordPressPlansPage(): Promise< void > {
+		await this.page.click( selectors.wpPlansContentOnlyOptionButton );
```
**question:non-blocking** - could you give a screenshot or link to the endpoint for this?

ouikhuan replied:
Please see the screenshot below, it's a pre-migration page with plans information in it. Users can click the content-only option for free import here. 
![Screen Shot 2023-05-29 at 11 22 58 PM](https://github.com/Automattic/wp-calypso/assets/4074459/a217b4ec-26ea-43c7-85c6-d53c03cce8cb)

worldomonation reviewed (changes requested):
Thanks for your PR.

I left a few questions, some of them blocking. If you aren't sure what any of those mean, do let one of us know!

Just a heads up that we're actively trying to discourage usage of E2E tests as the only form of solution. It sounds like this test can easily be implemented as a component test, doubly so because this E2E spec never actually imports any site. Has the team considered using the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/) to write component tests instead?

ouikhuan committed:
Change selector to role based selector and renaming functions

ouikhuan commented:
> Just a heads up that we're actively trying to discourage usage of E2E tests as the only form of solution. It sounds like this test can easily be implemented as a component test, doubly so because this E2E spec never actually imports any site. Has the team considered using the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/) to write component tests instead?

Thanks for the suggestion! Since our importer flow has changed due to the recent change, we need to update this existing e2e flow as well so it won't break the testing for now. We're planning to [add more tests](https://github.com/Automattic/wp-calypso/issues/77369) to the new flow soon and we can implement some component tests as well.

worldomonation commented on packages/calypso-e2e/src/lib/flows/start-import-flow.ts:
```diff
@@ -114,6 +115,13 @@ export class StartImportFlow {
+	 */
+	async validateWordPressPreMigrationPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.wpPreMigrationContentOnlyOptionButton );
+	}
```
**issues:blocking** - we are now recommending [developers use role-based locators](https://playwright.dev/docs/locators) unless there is no alternative. 

This is how you'd construct a locator:

```
await this.page.locator( selectors.wpPreMigrationContentOnlyOptionButton ).waitFor();
```

**suggestion:blocking** - since the `validateWordPressPreMigrationPage` and `contentOnlyWordPressPreMigrationPage` methods only differ in that one waits for a selector and one performs a click, is it really necessary to create two new methods for this? 

Since Playwright performs [automatic actionability checks prior to clicking](https://playwright.dev/docs/api/class-locator#locator-click), there is no need to do `waitForSelector` and a `click` in separate methods.

----

Taken together, the overall suggestion would be to:
1. combine the methods into one, and rename the method.
2. use a locator, not the deprecated `page.waitForSelector` method.

```suggestion
	async clickPremigrationOptionButton(): Promise< void > {
		await this.page.locator( selectors.wpPreMigrationContentOnlyOptionButton ).click();
	}
```

worldomonation commented on test/e2e/specs/onboarding/setup__importer.ts:
```diff
@@ -10,6 +10,7 @@ declare const browser: Browser;
 describe( DataHelper.createSuiteTitle( 'Site Import' ), () => {
 	const credentials = SecretsManager.secrets.testAccounts.defaultUser;
 
+	let redesignImporterFeature = false;
```
**suggestion:blocking** - in the interest of making sure temporary workarounds are removed later, could we please get a Tampermonkey-obfuscated link to the project thread here, with comments to remove this variable, the change in Line 23 and the if/else added to Line 48-54 once this redesigned importer is made default?

The same thing goes for the other file as well. Thanks!

worldomonation reviewed (changes requested):
Some blocking changes required, plus the previous comment on the actual need for more E2Es to consider. Thanks!

ouikhuan committed:
Combine two functions to one as suggested

ouikhuan committed:
Adding comments to remove variable after project goes live

ouikhuan commented:
> Some blocking changes required, plus the previous comment on the actual need for more E2Es to consider. Thanks!

Thanks for the detailed review and the suggestions @worldomonation ! I've updated the code based on your feedback, and also updated the tasks to include unit tests and component tests before adding more e2e tests.

worldomonation reviewed (approved):
Thanks for addressing the feedback thoroughly, I appreciate it!

worldomonation commented:
@ouikhuan This change has led to a huge spike in the flaky failure rate for the spec.

I'm going to strongly suggest that this PR be reverted for now to work out why the failure has spiked.
![image](https://github.com/Automattic/wp-calypso/assets/6549265/fef66f0c-abe6-4c4e-b672-0fcbef36c30c)
